### PR TITLE
Binary serialization support for context propagation and transport requests

### PR DESCRIPTION
**Pull request by AlexAPB**
This supports the serialization of an RPC (`transport.Request` with context propagation) as bytes. Redis/Kafka other queuing transports need to transfer the entire RPC to queue, to be picked up by some queue consumer.

### CODE REVIEW

**Comment by AlexAPB** [comment]
I'm not too fond of the naming for this package `marshal`, so willing to take suggestions

File: marshal/marshal.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	"go.uber.org/thriftrw/protocol"
+	"go.uber.org/thriftrw/wire"
+	"go.uber.org/yarpc/marshal/internal"
+	"go.uber.org/yarpc/transport"
+
+	"github.com/opentracing/opentracing-go"
+)
+
+// MarshalRPC encodes a SpanContext and Request into bytes
+func MarshalRPC(tracer opentracing.Tracer, spanContext opentracing.SpanContext, req *transport.Request) ([]byte, error) {
```
Comment (breerly) [inline marshal/marshal.go]: Use the package name to your advantage - suggest renaming `ToBytes`:

```
import marshal

marshal.ToBytes(...)
```

File: marshal/marshal.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		return nil, err
+	}
+
+	var writer bytes.Buffer
+	err = protocol.Binary.Encode(wireValue, &writer)
+	return writer.Bytes(), err
+}
+
+// UnmarshalRPC decodes bytes into a SpanContext and Request
+func UnmarshalRPC(tracer opentracing.Tracer, request []byte) (opentracing.SpanContext, *transport.Request, error) {
```
Comment (breerly) [inline marshal/marshal.go]: `FromBytes`

**Comment by breerly** [comment]
Also, I question whether this package should be a top-level package.

Who are the customers of these package? Aren't they transport implementors? Perhaps we should move this under the transport package, or directly in the transport package as:

```
import yarpc/transport

transport.MarshalToBytes()
transport.MarshalFromBytes()
```

File: transport/propagation.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+)
+
+// CreateOpentracingSpan creates a new context that has a reference to the
+// started span.
+// This should be called before a Outbound makes a call
+func CreateOpentracingSpan(
```
Comment (breerly) [inline transport/propagation.go]: This should be `CreateOpenTracingSpan`.

Source: http://opentracing.io/ spells it OpenTracing, not Opentracing.

File: serialize/serialize.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package serialize
```
Comment (kriskowal) [inline serialize/serialize.go]: serialize is a synonym for encode. maybe call this transport/envelope?

File: serialize/internal.thrift [hunk 5]
```diff
@@ -0,0 +1,14 @@
+struct RPC {
+	1: required binary spanContext
```
Comment (kriskowal) [inline serialize/internal.thrift]: I see: baggage carried here. Probably worth clarifying the binary encoding protocol for the span context here, with a reference to the opentracing binary carrier, and verifying that the wire encoding is stable @yurishkuro.

This presumably carries both baggage and tracing.

File: serialize/serialize.go (hunk 4 above)
Comment (breerly) [inline serialize/serialize.go]: @kriskowal I don't want to overload the term encode/encoding.

Comment (kriskowal) [inline serialize/serialize.go]: Agreed, ergo "transport/envelope".

Comment (AlexAPB) [inline serialize/serialize.go]: I was talking to @breerly / @willhug about this. Similar to moving Peer to the top level directory, I didn't want to cloud the transport directory with a package that's not a transport

File: serialize/internal.thrift (hunk 5 above)
Comment (yurishkuro) [inline serialize/internal.thrift]: > This presumably carries both baggage and tracing.

It does. But we have never used it with Jaeger clients, I think it's not implemented in most languages, so would need testing & extra dev. Why not use `map<string,string>`?

**Pull request merged**